source/particle/world.cc

```diff
@@ -0,0 +1,889 @@
+                      const Point<dim> p_unit = this->get_mapping().transform_real_to_unit_cell(child, p.get_location());
+                      if (GeometryInfo<dim>::is_inside_unit_cell(p_unit))
+                        {
+                          particles.insert(std::make_pair(std::make_pair(child->level(),child->index()),p));
+                        }
+                    }
+                  catch (...)
+                    {}
+                }
+            }
+        }
+
+      if (particle_load_balancing == remove_and_add_particles)
+        {
+          // TODO: Add particles
+        }
+    }
+
+    template <int dim>
+    void
+    World<dim>::sort_particles_in_subdomains_and_cells()
+    {
+      // There are three reasons why a particle is not in its old cell:
+      // It moved to another cell, to another domain or it left the mesh.
+      // Sort the particles accordingly and deal with them
+      std::multimap<types::LevelInd, Particle<dim> >            moved_particles_cell;
+      std::multimap<types::subdomain_id, Particle<dim> > moved_particles_domain;
+      std::multimap<types::LevelInd, Particle<dim> >            lost_particles;
+
+      // Find the cells that the particles moved to.
+      // Note that the iterator in the following loop is increased in a
+      // very particular way, because it is changed, if elements
+      // get erased. A change can result in invalid memory access.
+      typename std::multimap<types::LevelInd, Particle<dim> >::iterator   it;
+      for (it=particles.begin(); it!=particles.end();)
+        {
+          // If we know the particles old cell, check if it is still inside
+          if (it->first != std::make_pair(-1,-1))
+            {
+              const typename parallel::distributed::Triangulation<dim>::active_cell_iterator
+              old_cell (&(this->get_triangulation()), it->first.first, it->first.second);
+
+              try
+                {
+                  const Point<dim> p_unit = this->get_mapping().transform_real_to_unit_cell(old_cell, it->second.get_location());
+                  if (GeometryInfo<dim>::is_inside_unit_cell(p_unit))
+                    {
+                      ++it;
+                      continue;
+                    }
+                }
+              catch (...)
+                {}
+            }
+
+          // The particle is not in its old cell. Look for the new cell.
+          typename parallel::distributed::Triangulation<dim>::active_cell_iterator cell;
+          try
+            {
+              cell = (GridTools::find_active_cell_around_point<> (this->get_mapping(), this->get_triangulation(), it->second.get_location())).first;
+            }
+          catch (GridTools::ExcPointNotFound<dim> &)
+            {
+              // If we can find no cell for this particle it has left the domain due
+              // to an integration error or an open boundary.
+              lost_particles.insert(*it);
+              particles.erase(it++);
```

Comment:
The fact that you need an explanation why it works suggests that the code might be hard to maintain in the long run because we forget. How about something like this:
```
  // move 'it' to the next particle, then remove the element of the map it previously pointed to
  std::map<...>::iterator tmp = it;
  ++it;
  particles.erase(tmp);
```